services/mac_manager/src/main/java/com/futurewei/alcor/macmanager/service/implement/MacRedisServiceImpl.java

```diff
@@ -61,7 +64,48 @@ public MacState createMacState(MacState macState) throws Exception {
         MacAddress macAddress = new MacAddress();
         if (macState.getState() == null)
             macState.setState(MacUtil.MAC_STATE_ACTIVE);
-        else if (macState.getState().trim().length()==0)
+        else if (macState.getState().trim().length() == 0)
+            macState.setState(MacUtil.MAC_STATE_ACTIVE);
+        if (macPoolRedisRepository.getSize() < (nMacPoolSize - 10)) {
+            CompletableFuture<Long> completableFuture = CompletableFuture.supplyAsync(() -> {
+                long n = 0;
+                try {
+                    n = generateMacInPool(20);
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+                return n;
+            });
+            long l = completableFuture.get();
+            completableFuture.thenAccept(System.out::println);
+            System.out.println(("# of new MAC: " + l));
+            completableFuture.join();
+        }
+
+        String strMacAddress = allocateMacState(macState);
```

Comment:
Another issue, when the mac range is full, the current algorithm could retry unlimited times. We will need to impose a limit on the number of retry.

Reply:
(1) Yes, mac_allocation does 2 DB transactions to prevent duplication. (delete from mac_pool, check mac_stat to check duplication).
(2) I will make this configurable now. Later if there is a better solution, it will be applied.

Reply:
The current Redis impl doesn’t support transactions. See how a transaction in ip manager is implemented and used.

Reply:
All DAO methods synchronized.

Reply:
synchronized wouldn't solve the problem as synchronized keyword is used to handle multiple threads inside a MacManager instance. We were talking about multiple MacManager instances.